pyscriptjs/src/python/pyscript.py

```diff
@@ -490,4 +489,139 @@ def wrapper(class_):
+    name.
+    """
+
+    def __init__(self, name, obj, message):
+        self.__name = name
+        self.__obj = obj
+        self.__message = message
+        self.__warning_already_shown = False
+
+    def __repr__(self):
+        return f"<DeprecatedGlobal({self.__name!r})>"
+
+    def _show_warning(self, message):
+        """
+        NOTE: this is overridden by unit tests
+        """
+        # this showWarning is implemented in js and injected into this
+        # namespace by main.ts
+        showWarning(message, "html")  # noqa: F821
+
+    def _show_warning_maybe(self):
+        if self.__warning_already_shown:
+            return
+        self._show_warning(self.__message)
+        self.__warning_already_shown = True
+
+    def __getattr__(self, attr):
+        self._show_warning_maybe()
+        return getattr(self.__obj, attr)
+
+    def __call__(self, *args, **kwargs):
+        self._show_warning_maybe()
+        return self.__obj(*args, **kwargs)
+
+
+class PyScript:
+    """
+    This class is deprecated since 2022.12.1.
+
+    All its old functionalities are available as module-level functions. This
+    class should be killed eventually.
+    """
+
+    loop = loop
+
+    @staticmethod
+    def run_until_complete(f):
+        run_until_complete(f)
+
+    @staticmethod
+    def write(element_id, value, append=False, exec_id=0):
+        write(element_id, value, append, exec_id)
+
+
+def _install_deprecated_globals_2022_12_1(ns):
+    """
+    Install into the given namespace all the globals which have been
+    deprecated since the 2022.12.1 release. Eventually they should be killed.
+    """
+
+    def deprecate(name, obj, instead):
+        message = f"Direct usage of <code>{name}</code> is deprecated. " + instead
+        ns[name] = DeprecatedGlobal(name, obj, message)
+
+    # function/classes defined in pyscript.py ===> pyscript.XXX
+    pyscript_names = [
+        "PyItemTemplate",
+        "PyListTemplate",
+        "PyWidgetTheme",
+        "add_classes",
+        "create",
+        "loop",
+    ]
+    for name in pyscript_names:
+        deprecate(
+            name, globals()[name], f"Please use <code>pyscript.{name}</code> instead."
+        )
+
+    # stdlib modules ===> import XXX
+    stdlib_names = [
+        "asyncio",
+        "base64",
+        "io",
+        "sys",
+        "time",
+        "datetime",
+        "pyodide",
+        "micropip",
+    ]
+    for name in stdlib_names:
+        obj = __import__(name)
+        deprecate(name, obj, f"Please use <code>import {name}</code> instead.")
+
+    # special case
+    deprecate(
+        "dedent", dedent, "Please use <code>from textwrap import dedent</code> instead."
+    )
+
+    # these are names that used to leak in the globals but they are just
+    # implementation details. People should not use them.
+    private_names = ["eval_formatter", "format_mime", "identity", "render_image"]
```

Comment:
> in previous discussions, some users have been routed to look at these attributes

ouch, I wasn't aware of this :man_facepalming: 

Adding them is not straightforward though, because the current `DeprecatedGlobal` is not enough: you would need a `DeprecatedDictGlobal` or something like that.
Is it worth? (Genuine question, I don't really know).

This is the problems with polluting the global namespace: because of how python works, it's impossible to deprecate them "cleanly"

Reply:
I correct myself, I found a way to support `MIME_METHODS` and `MIME_RENDERERS` easily. Fixed by ef80b7b

Reply:
Very nice!